envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/utils/GzipUtils.kt

```diff
@@ -0,0 +1,42 @@
+package pl.allegro.tech.servicemesh.envoycontrol.utils
+
+import com.fasterxml.jackson.core.type.TypeReference
+import com.fasterxml.jackson.databind.ObjectMapper
+import pl.allegro.tech.servicemesh.envoycontrol.logger
+import java.io.ByteArrayOutputStream
+import java.lang.reflect.Type
+import java.util.zip.GZIPInputStream
+import java.util.zip.GZIPOutputStream
+import kotlin.text.Charsets.UTF_8
+
+class GzipUtils(val objectMapper: ObjectMapper) {
+
+    private val logger by logger()
+
+    fun gzip(content: Any): ByteArray {
+        try {
+            val json = objectMapper.writeValueAsString(content)
+            val outputStream = ByteArrayOutputStream()
+            GZIPOutputStream(outputStream).bufferedWriter(UTF_8).use { it.write(json) }
+            return outputStream.toByteArray()
+        } catch (e: Exception) {
+            logger.error(e.message)
+            throw IllegalStateException("Couldn't compress data $content")
+        }
+    }
+
+    fun <T> unGzip(content: ByteArray, javaClass: Class<T>): T {
```

Comment:
I think that you can create ClientHttpRequestInterceptor with decoding logic and add it to restTemplate. Thanks that we can omit creating different restTemplates  depending on `gzip` property

Reply:
hmhmh I found out that HttpComponentsClientHttpRequestFactory has out of the box support for un-gzipping requests, so I switched to it in rest template config

Reply:
IMO, it would be better to create one interceptor, than adding additional dependency - also there are some implementations of such an interceptor in our allegro-internal's repositories .

Reply:
added interceptor